calculator_test.go

```diff
@@ -3,189 +3,427 @@ package calculator_test
 import (
 	"calculator"
 	"math"
+	"math/rand"
 	"testing"
+	"time"
 )
 
+func closeEnough(a, b float64) bool {
+	const epsilon = 1e-9
+	return math.Abs(a-b) <= epsilon
+}
+
 func TestAdd(t *testing.T) {
 	t.Parallel()
-	var want float64 = 4
-	got := calculator.Add(2, 2)
-	if want != got {
-		t.Errorf("want %f, got %f", want, got)
+	tcs := []struct {
+		name string
+		a    float64
+		b    float64
+		nums []float64
+		want float64
+	}{
+		{
+			name: "Two positive numbers",
+			a:    3,
+			b:    2,
+			want: 5,
+		},
+		{
+			name: "Zero subtracted with any number",
+			a:    0,
+			b:    -2,
+			want: -2,
+		},
+		{
+			name: "Positive and negative number",
+			a:    -5,
+			b:    4,
+			want: -1,
+		},
+		{
+			name: "Two negative numbers",
+			a:    -8,
+			b:    -3,
+			want: -11,
+		},
+		{
+			name: "Two fractional numbers",
+			a:    3.3,
+			b:    3.2,
+			want: 6.5,
+		},
+		{
+			name: "More than two negative numbers",
+			a:    -8,
+			b:    -3,
+			nums: []float64{-1, -2, -3},
+			want: -17,
+		},
+
+		{
+			name: "More than two fractional numbers",
+			a:    3.3,
+			b:    3.2,
+			nums: []float64{2.1},
+			want: 8.6,
+		},
 	}
-}
 
+	for _, tc := range tcs {
+		got := calculator.Add(tc.a, tc.b, tc.nums...)
+		if !closeEnough(tc.want, got) {
+			t.Errorf("%s : Add(%f, %f, %v) : want %f, got %f",
+				tc.name, tc.a, tc.b, tc.nums, tc.want, got)
+		}
+	}
+}
 
-func almostEqual(a, b float64) bool {
-	const float64EqualityThreshold = 1e-9
-	return math.Abs(a-b) <= float64EqualityThreshold
+func TestAddRandom(t *testing.T) {
+	t.Parallel()
+	rand.Seed(time.Now().UnixNano())
+	for i := 1; i <= 100; i++ {
```

Comment:
Since 99% of all `for` loops start at zero (citation needed) I would start this one at zero too! I think of this as the 'Principle of Least Surprise' for the reader.

Reply:
Yeah it should from 0, there's no other explanation than maybe a residual of some experimentation. Thanks for spotting it.